pkg/dataobj/internal/dataset/page.go

```diff
@@ -57,11 +69,52 @@ type Page interface {
 	ReadPage(ctx context.Context) (PageData, error)
 }
 
+type byteBuffer struct {
+	data []byte
+}
+
+// Resize changes the len of buf to n. If n is greater than the
+// capacity of buf, Resize allocates a new slice.
+func (buf *byteBuffer) Resize(n int) {
+	if n > len(buf.data) {
+		tmp := make([]byte, n)
+		copy(buf.data, tmp)
+		buf.data = tmp
+	} else {
+		buf.data = buf.data[:n]
+	}
+}
+
+// Bytes returns the underlying data of buf.
+func (buf *byteBuffer) Bytes() []byte {
+	return buf.data
+}
+
+var bufferPool = sync.Pool{
+	New: func() any {
+		return new(byteBuffer)
+	},
+}
+
+type releasableData struct {
+	buf *byteBuffer
+}
+
+func (rd *releasableData) Bytes() []byte { return rd.buf.Bytes() }
+
+func (rd *releasableData) Close() error {
+	if rd.buf != nil {
+		bufferPool.Put(rd.buf)
+		rd.buf = nil
+	}
+	return nil
+}
+
 // MemPage holds an encoded (and optionally compressed) sequence of [Value]
 // entries of a common type. Use [ColumnBuilder] to construct sets of pages.
 type MemPage struct {
-	Desc PageDesc // Description of the page.
-	Data PageData // Data for the page.
+	Desc PageDesc        // Description of the page.
+	data *releasableData // Data for the page.
```

Comment:
I'm a little worried about MemPage being releasable. 

When a column is constructed through a [ColumnBuilder](https://github.com/grafana/loki/blob/20e04a4192fe1f8eacebc4ac241a3f293d328e1b/pkg/dataobj/internal/dataset/column_builder.go#L180), it returns a MemColumn, which is a wrapper around MemPages. 

MemColumn and MemPages implement the proper interfaces to allow you to read them using a Reader, but MemPages being releasable now means that you can only read them exactly once before you lose access to the memory. 

I don't think there's any current case where we read a MemColumn more than once, but it feels risky enough that I don't think MemPage should be using releasable data.